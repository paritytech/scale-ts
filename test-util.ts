/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
