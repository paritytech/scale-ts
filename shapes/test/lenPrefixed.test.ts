import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

testShape($.lenPrefixed($.constant(null)), [null])

testShape($.lenPrefixed($.array($.compact($.u32))), {
  empty: [],
  primes: [2, 3, 5, 7, 11, 13, 17, 23, 29],
  squares: [...Array(1000)].map((_, i) => i ** 2),
})

testShape($.lenPrefixed($.promise($.compact($.u32))), {
  123: () => new Promise((r) => setTimeout(() => r(123), 100)),
}, true)

testInvalid($.lenPrefixed($.u8), [null, undefined, -1])
