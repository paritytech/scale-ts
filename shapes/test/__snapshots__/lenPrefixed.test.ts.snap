export const snapshot = {};

snapshot[`\$.lenPrefixed(\$.constant(null)) null 1`] = `00`;

snapshot[`\$.lenPrefixed(\$.array(\$.compact(\$.u32))) empty 1`] = `
04
00
`;

snapshot[`\$.lenPrefixed(\$.array(\$.compact(\$.u32))) primes 1`] = `
28
24
08
0c
14
1c
2c
34
44
5c
74
`;

snapshot[`\$.lenPrefixed(\$.array(\$.compact(\$.u32))) squares 1`] = `
69
3a
a1
0f
00
04
10
24
40
64
90
c4
01
01
45
01
91
01
e5
01
41
02
a5
02
11
03
85
03
01
04
85
04
11
05
a5
05
41
06
e5
06
91
07
45
08
01
09
c5
09
91
0a
65
0b
41
0c
25
0d
11
0e
05
0f
01
10
05
11
11
12
25
13
41
14
65
15
91
16
c5
17
01
19
45
1a
91
1b
e5
1c
41
1e
a5
1f
11
21
85
22
01
24
85
25
11
27
a5
28
41
2a
e5
2b
91
2d
45
2f
01
31
c5
32
91
34
65
36
41
38
25
3a
11
3c
05
3e
01
40
05
42
11
44
25
46
41
48
65
4a
91
4c
c5
4e
01
51
45
53
91
55
e5
57
41
5a
a5
5c
11
5f
85
61
01
64
85
66
11
69
a5
6b
41
6e
e5
70
91
73
45
76
01
79
c5
7b
91
7e
65
81
41
84
25
87
11
8a
05
8d
01
90
05
93
11
96
25
99
41
9c
65
9f
91
a2
c5
a5
01
a9
45
ac
91
af
e5
b2
41
b6
a5
b9
11
bd
85
c0
01
c4
85
c7
11
cb
a5
ce
41
d2
e5
d5
91
d9
45
dd
01
e1
c5
e4
91
e8
65
ec
41
f0
25
f4
11
f8
05
fc
02
00
01
00
06
04
01
00
12
08
01
00
26
0c
01
00
42
10
01
00
66
14
01
00
92
18
01
00
c6
1c
01
00
02
21
01
00
46
25
01
00
92
29
01
00
e6
2d
01
00
42
32
01
00
a6
36
01
00
12
3b
01
00
86
3f
01
00
02
44
01
00
86
48
01
00
12
4d
01
00
a6
51
01
00
42
56
01
00
e6
5a
01
00
92
5f
01
00
46
64
01
00
02
69
01
00
c6
6d
01
00
92
72
01
00
66
77
01
00
42
7c
01
00
26
81
01
00
12
86
01
00
06
8b
01
00
02
90
01
00
06
95
01
00
12
9a
01
00
26
9f
01
00
42
a4
01
00
66
a9
01
00
92
ae
01
00
c6
b3
01
00
02
b9
01
00
46
be
01
00
92
c3
01
00
e6
c8
01
00
42
ce
01
00
a6
d3
01
00
12
d9
01
00
86
de
01
00
02
e4
01
00
86
e9
01
00
12
ef
01
00
a6
f4
01
00
42
fa
01
00
e6
ff
01
00
92
05
02
00
46
0b
02
00
02
11
02
00
c6
16
02
00
92
1c
02
00
66
22
02
00
42
28
02
00
26
2e
02
00
12
34
02
00
06
3a
02
00
02
40
02
00
06
46
02
00
12
4c
02
00
26
52
02
00
42
58
02
00
66
5e
02
00
92
64
02
00
c6
6a
02
00
02
71
02
00
46
77
02
00
92
7d
02
00
e6
83
02
00
42
8a
02
00
a6
90
02
00
12
97
02
00
86
9d
02
00
02
a4
02
00
86
aa
02
00
12
b1
02
00
a6
b7
02
00
42
be
02
00
e6
c4
02
00
92
cb
02
00
46
d2
02
00
02
d9
02
00
c6
df
02
00
92
e6
02
00
66
ed
02
00
42
f4
02
00
26
fb
02
00
12
02
03
00
06
09
03
00
02
10
03
00
06
17
03
00
12
1e
03
00
26
25
03
00
42
2c
03
00
66
33
03
00
92
3a
03
00
c6
41
03
00
02
49
03
00
46
50
03
00
92
57
03
00
e6
5e
03
00
42
66
03
00
a6
6d
03
00
12
75
03
00
86
7c
03
00
02
84
03
00
86
8b
03
00
12
93
03
00
a6
9a
03
00
42
a2
03
00
e6
a9
03
00
92
b1
03
00
46
b9
03
00
02
c1
03
00
c6
c8
03
00
92
d0
03
00
66
d8
03
00
42
e0
03
00
26
e8
03
00
12
f0
03
00
06
f8
03
00
02
00
04
00
06
08
04
00
12
10
04
00
26
18
04
00
42
20
04
00
66
28
04
00
92
30
04
00
c6
38
04
00
02
41
04
00
46
49
04
00
92
51
04
00
e6
59
04
00
42
62
04
00
a6
6a
04
00
12
73
04
00
86
7b
04
00
02
84
04
00
86
8c
04
00
12
95
04
00
a6
9d
04
00
42
a6
04
00
e6
ae
04
00
92
b7
04
00
46
c0
04
00
02
c9
04
00
c6
d1
04
00
92
da
04
00
66
e3
04
00
42
ec
04
00
26
f5
04
00
12
fe
04
00
06
07
05
00
02
10
05
00
06
19
05
00
12
22
05
00
26
2b
05
00
42
34
05
00
66
3d
05
00
92
46
05
00
c6
4f
05
00
02
59
05
00
46
62
05
00
92
6b
05
00
e6
74
05
00
42
7e
05
00
a6
87
05
00
12
91
05
00
86
9a
05
00
02
a4
05
00
86
ad
05
00
12
b7
05
00
a6
c0
05
00
42
ca
05
00
e6
d3
05
00
92
dd
05
00
46
e7
05
00
02
f1
05
00
c6
fa
05
00
92
04
06
00
66
0e
06
00
42
18
06
00
26
22
06
00
12
2c
06
00
06
36
06
00
02
40
06
00
06
4a
06
00
12
54
06
00
26
5e
06
00
42
68
06
00
66
72
06
00
92
7c
06
00
c6
86
06
00
02
91
06
00
46
9b
06
00
92
a5
06
00
e6
af
06
00
42
ba
06
00
a6
c4
06
00
12
cf
06
00
86
d9
06
00
02
e4
06
00
86
ee
06
00
12
f9
06
00
a6
03
07
00
42
0e
07
00
e6
18
07
00
92
23
07
00
46
2e
07
00
02
39
07
00
c6
43
07
00
92
4e
07
00
66
59
07
00
42
64
07
00
26
6f
07
00
12
7a
07
00
06
85
07
00
02
90
07
00
06
9b
07
00
12
a6
07
00
26
b1
07
00
42
bc
07
00
66
c7
07
00
92
d2
07
00
c6
dd
07
00
02
e9
07
00
46
f4
07
00
92
ff
07
00
e6
0a
08
00
42
16
08
00
a6
21
08
00
12
2d
08
00
86
38
08
00
02
44
08
00
86
4f
08
00
12
5b
08
00
a6
66
08
00
42
72
08
00
e6
7d
08
00
92
89
08
00
46
95
08
00
02
a1
08
00
c6
ac
08
00
92
b8
08
00
66
c4
08
00
42
d0
08
00
26
dc
08
00
12
e8
08
00
06
f4
08
00
02
00
09
00
06
0c
09
00
12
18
09
00
26
24
09
00
42
30
09
00
66
3c
09
00
92
48
09
00
c6
54
09
00
02
61
09
00
46
6d
09
00
92
79
09
00
e6
85
09
00
42
92
09
00
a6
9e
09
00
12
ab
09
00
86
b7
09
00
02
c4
09
00
86
d0
09
00
12
dd
09
00
a6
e9
09
00
42
f6
09
00
e6
02
0a
00
92
0f
0a
00
46
1c
0a
00
02
29
0a
00
c6
35
0a
00
92
42
0a
00
66
4f
0a
00
42
5c
0a
00
26
69
0a
00
12
76
0a
00
06
83
0a
00
02
90
0a
00
06
9d
0a
00
12
aa
0a
00
26
b7
0a
00
42
c4
0a
00
66
d1
0a
00
92
de
0a
00
c6
eb
0a
00
02
f9
0a
00
46
06
0b
00
92
13
0b
00
e6
20
0b
00
42
2e
0b
00
a6
3b
0b
00
12
49
0b
00
86
56
0b
00
02
64
0b
00
86
71
0b
00
12
7f
0b
00
a6
8c
0b
00
42
9a
0b
00
e6
a7
0b
00
92
b5
0b
00
46
c3
0b
00
02
d1
0b
00
c6
de
0b
00
92
ec
0b
00
66
fa
0b
00
42
08
0c
00
26
16
0c
00
12
24
0c
00
06
32
0c
00
02
40
0c
00
06
4e
0c
00
12
5c
0c
00
26
6a
0c
00
42
78
0c
00
66
86
0c
00
92
94
0c
00
c6
a2
0c
00
02
b1
0c
00
46
bf
0c
00
92
cd
0c
00
e6
db
0c
00
42
ea
0c
00
a6
f8
0c
00
12
07
0d
00
86
15
0d
00
02
24
0d
00
86
32
0d
00
12
41
0d
00
a6
4f
0d
00
42
5e
0d
00
e6
6c
0d
00
92
7b
0d
00
46
8a
0d
00
02
99
0d
00
c6
a7
0d
00
92
b6
0d
00
66
c5
0d
00
42
d4
0d
00
26
e3
0d
00
12
f2
0d
00
06
01
0e
00
02
10
0e
00
06
1f
0e
00
12
2e
0e
00
26
3d
0e
00
42
4c
0e
00
66
5b
0e
00
92
6a
0e
00
c6
79
0e
00
02
89
0e
00
46
98
0e
00
92
a7
0e
00
e6
b6
0e
00
42
c6
0e
00
a6
d5
0e
00
12
e5
0e
00
86
f4
0e
00
02
04
0f
00
86
13
0f
00
12
23
0f
00
a6
32
0f
00
42
42
0f
00
e6
51
0f
00
92
61
0f
00
46
71
0f
00
02
81
0f
00
c6
90
0f
00
92
a0
0f
00
66
b0
0f
00
42
c0
0f
00
26
d0
0f
00
12
e0
0f
00
06
f0
0f
00
02
00
10
00
06
10
10
00
12
20
10
00
26
30
10
00
42
40
10
00
66
50
10
00
92
60
10
00
c6
70
10
00
02
81
10
00
46
91
10
00
92
a1
10
00
e6
b1
10
00
42
c2
10
00
a6
d2
10
00
12
e3
10
00
86
f3
10
00
02
04
11
00
86
14
11
00
12
25
11
00
a6
35
11
00
42
46
11
00
e6
56
11
00
92
67
11
00
46
78
11
00
02
89
11
00
c6
99
11
00
92
aa
11
00
66
bb
11
00
42
cc
11
00
26
dd
11
00
12
ee
11
00
06
ff
11
00
02
10
12
00
06
21
12
00
12
32
12
00
26
43
12
00
42
54
12
00
66
65
12
00
92
76
12
00
c6
87
12
00
02
99
12
00
46
aa
12
00
92
bb
12
00
e6
cc
12
00
42
de
12
00
a6
ef
12
00
12
01
13
00
86
12
13
00
02
24
13
00
86
35
13
00
12
47
13
00
a6
58
13
00
42
6a
13
00
e6
7b
13
00
92
8d
13
00
46
9f
13
00
02
b1
13
00
c6
c2
13
00
92
d4
13
00
66
e6
13
00
42
f8
13
00
26
0a
14
00
12
1c
14
00
06
2e
14
00
02
40
14
00
06
52
14
00
12
64
14
00
26
76
14
00
42
88
14
00
66
9a
14
00
92
ac
14
00
c6
be
14
00
02
d1
14
00
46
e3
14
00
92
f5
14
00
e6
07
15
00
42
1a
15
00
a6
2c
15
00
12
3f
15
00
86
51
15
00
02
64
15
00
86
76
15
00
12
89
15
00
a6
9b
15
00
42
ae
15
00
e6
c0
15
00
92
d3
15
00
46
e6
15
00
02
f9
15
00
c6
0b
16
00
92
1e
16
00
66
31
16
00
42
44
16
00
26
57
16
00
12
6a
16
00
06
7d
16
00
02
90
16
00
06
a3
16
00
12
b6
16
00
26
c9
16
00
42
dc
16
00
66
ef
16
00
92
02
17
00
c6
15
17
00
02
29
17
00
46
3c
17
00
92
4f
17
00
e6
62
17
00
42
76
17
00
a6
89
17
00
12
9d
17
00
86
b0
17
00
02
c4
17
00
86
d7
17
00
12
eb
17
00
a6
fe
17
00
42
12
18
00
e6
25
18
00
92
39
18
00
46
4d
18
00
02
61
18
00
c6
74
18
00
92
88
18
00
66
9c
18
00
42
b0
18
00
26
c4
18
00
12
d8
18
00
06
ec
18
00
02
00
19
00
06
14
19
00
12
28
19
00
26
3c
19
00
42
50
19
00
66
64
19
00
92
78
19
00
c6
8c
19
00
02
a1
19
00
46
b5
19
00
92
c9
19
00
e6
dd
19
00
42
f2
19
00
a6
06
1a
00
12
1b
1a
00
86
2f
1a
00
02
44
1a
00
86
58
1a
00
12
6d
1a
00
a6
81
1a
00
42
96
1a
00
e6
aa
1a
00
92
bf
1a
00
46
d4
1a
00
02
e9
1a
00
c6
fd
1a
00
92
12
1b
00
66
27
1b
00
42
3c
1b
00
26
51
1b
00
12
66
1b
00
06
7b
1b
00
02
90
1b
00
06
a5
1b
00
12
ba
1b
00
26
cf
1b
00
42
e4
1b
00
66
f9
1b
00
92
0e
1c
00
c6
23
1c
00
02
39
1c
00
46
4e
1c
00
92
63
1c
00
e6
78
1c
00
42
8e
1c
00
a6
a3
1c
00
12
b9
1c
00
86
ce
1c
00
02
e4
1c
00
86
f9
1c
00
12
0f
1d
00
a6
24
1d
00
42
3a
1d
00
e6
4f
1d
00
92
65
1d
00
46
7b
1d
00
02
91
1d
00
c6
a6
1d
00
92
bc
1d
00
66
d2
1d
00
42
e8
1d
00
26
fe
1d
00
12
14
1e
00
06
2a
1e
00
02
40
1e
00
06
56
1e
00
12
6c
1e
00
26
82
1e
00
42
98
1e
00
66
ae
1e
00
92
c4
1e
00
c6
da
1e
00
02
f1
1e
00
46
07
1f
00
92
1d
1f
00
e6
33
1f
00
42
4a
1f
00
a6
60
1f
00
12
77
1f
00
86
8d
1f
00
02
a4
1f
00
86
ba
1f
00
12
d1
1f
00
a6
e7
1f
00
42
fe
1f
00
e6
14
20
00
92
2b
20
00
46
42
20
00
02
59
20
00
c6
6f
20
00
92
86
20
00
66
9d
20
00
42
b4
20
00
26
cb
20
00
12
e2
20
00
06
f9
20
00
02
10
21
00
06
27
21
00
12
3e
21
00
26
55
21
00
42
6c
21
00
66
83
21
00
92
9a
21
00
c6
b1
21
00
02
c9
21
00
46
e0
21
00
92
f7
21
00
e6
0e
22
00
42
26
22
00
a6
3d
22
00
12
55
22
00
86
6c
22
00
02
84
22
00
86
9b
22
00
12
b3
22
00
a6
ca
22
00
42
e2
22
00
e6
f9
22
00
92
11
23
00
46
29
23
00
02
41
23
00
c6
58
23
00
92
70
23
00
66
88
23
00
42
a0
23
00
26
b8
23
00
12
d0
23
00
06
e8
23
00
02
00
24
00
06
18
24
00
12
30
24
00
26
48
24
00
42
60
24
00
66
78
24
00
92
90
24
00
c6
a8
24
00
02
c1
24
00
46
d9
24
00
92
f1
24
00
e6
09
25
00
42
22
25
00
a6
3a
25
00
12
53
25
00
86
6b
25
00
02
84
25
00
86
9c
25
00
12
b5
25
00
a6
cd
25
00
42
e6
25
00
e6
fe
25
00
92
17
26
00
46
30
26
00
02
49
26
00
c6
61
26
00
92
7a
26
00
66
93
26
00
42
ac
26
00
26
c5
26
00
12
de
26
00
06
f7
26
00
02
10
27
00
06
29
27
00
12
42
27
00
26
5b
27
00
42
74
27
00
66
8d
27
00
92
a6
27
00
c6
bf
27
00
02
d9
27
00
46
f2
27
00
92
0b
28
00
e6
24
28
00
42
3e
28
00
a6
57
28
00
12
71
28
00
86
8a
28
00
02
a4
28
00
86
bd
28
00
12
d7
28
00
a6
f0
28
00
42
0a
29
00
e6
23
29
00
92
3d
29
00
46
57
29
00
02
71
29
00
c6
8a
29
00
92
a4
29
00
66
be
29
00
42
d8
29
00
26
f2
29
00
12
0c
2a
00
06
26
2a
00
02
40
2a
00
06
5a
2a
00
12
74
2a
00
26
8e
2a
00
42
a8
2a
00
66
c2
2a
00
92
dc
2a
00
c6
f6
2a
00
02
11
2b
00
46
2b
2b
00
92
45
2b
00
e6
5f
2b
00
42
7a
2b
00
a6
94
2b
00
12
af
2b
00
86
c9
2b
00
02
e4
2b
00
86
fe
2b
00
12
19
2c
00
a6
33
2c
00
42
4e
2c
00
e6
68
2c
00
92
83
2c
00
46
9e
2c
00
02
b9
2c
00
c6
d3
2c
00
92
ee
2c
00
66
09
2d
00
42
24
2d
00
26
3f
2d
00
12
5a
2d
00
06
75
2d
00
02
90
2d
00
06
ab
2d
00
12
c6
2d
00
26
e1
2d
00
42
fc
2d
00
66
17
2e
00
92
32
2e
00
c6
4d
2e
00
02
69
2e
00
46
84
2e
00
92
9f
2e
00
e6
ba
2e
00
42
d6
2e
00
a6
f1
2e
00
12
0d
2f
00
86
28
2f
00
02
44
2f
00
86
5f
2f
00
12
7b
2f
00
a6
96
2f
00
42
b2
2f
00
e6
cd
2f
00
92
e9
2f
00
46
05
30
00
02
21
30
00
c6
3c
30
00
92
58
30
00
66
74
30
00
42
90
30
00
26
ac
30
00
12
c8
30
00
06
e4
30
00
02
00
31
00
06
1c
31
00
12
38
31
00
26
54
31
00
42
70
31
00
66
8c
31
00
92
a8
31
00
c6
c4
31
00
02
e1
31
00
46
fd
31
00
92
19
32
00
e6
35
32
00
42
52
32
00
a6
6e
32
00
12
8b
32
00
86
a7
32
00
02
c4
32
00
86
e0
32
00
12
fd
32
00
a6
19
33
00
42
36
33
00
e6
52
33
00
92
6f
33
00
46
8c
33
00
02
a9
33
00
c6
c5
33
00
92
e2
33
00
66
ff
33
00
42
1c
34
00
26
39
34
00
12
56
34
00
06
73
34
00
02
90
34
00
06
ad
34
00
12
ca
34
00
26
e7
34
00
42
04
35
00
66
21
35
00
92
3e
35
00
c6
5b
35
00
02
79
35
00
46
96
35
00
92
b3
35
00
e6
d0
35
00
42
ee
35
00
a6
0b
36
00
12
29
36
00
86
46
36
00
02
64
36
00
86
81
36
00
12
9f
36
00
a6
bc
36
00
42
da
36
00
e6
f7
36
00
92
15
37
00
46
33
37
00
02
51
37
00
c6
6e
37
00
92
8c
37
00
66
aa
37
00
42
c8
37
00
26
e6
37
00
12
04
38
00
06
22
38
00
02
40
38
00
06
5e
38
00
12
7c
38
00
26
9a
38
00
42
b8
38
00
66
d6
38
00
92
f4
38
00
c6
12
39
00
02
31
39
00
46
4f
39
00
92
6d
39
00
e6
8b
39
00
42
aa
39
00
a6
c8
39
00
12
e7
39
00
86
05
3a
00
02
24
3a
00
86
42
3a
00
12
61
3a
00
a6
7f
3a
00
42
9e
3a
00
e6
bc
3a
00
92
db
3a
00
46
fa
3a
00
02
19
3b
00
c6
37
3b
00
92
56
3b
00
66
75
3b
00
42
94
3b
00
26
b3
3b
00
12
d2
3b
00
06
f1
3b
00
02
10
3c
00
06
2f
3c
00
12
4e
3c
00
26
6d
3c
00
42
8c
3c
00
66
ab
3c
00
92
ca
3c
00
c6
e9
3c
00
`;

snapshot[`\$.lenPrefixed(\$.promise(\$.compact(\$.u32))) 123 1`] = `
08
ed
01
`;

snapshot[`\$.lenPrefixed(\$.u8) invalid null 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.lenPrefixed(\$.u8) invalid undefined 1`] = `ShapeAssertError: typeof value !== "number"`;

snapshot[`\$.lenPrefixed(\$.u8) invalid -1 1`] = `ShapeAssertError: value < 0`;
